### PR TITLE
New methods added

### PR DESCRIPTION
**Pull request by maxcapodi78**
- Maxwell 3d will have a get_conduction_paths method starting from AEDT 23R1
- hfss3dlayout components has multiple methods to enable die and solderballs in IC, IO and Other components type
- hfss3dlayout has now a method to get the x_t generated during mesh phase (23R1)
- hfss3dlayout.mesh has a generate_mesh command

### CODE REVIEW

File: pyaedt/hfss3dlayout.py [hunk 1]
```diff
@@ -356,7 +357,7 @@ def create_coax_port(self, vianame, radial_extent, layer, alignment="lower"):
         radial_extent : float
             Radial coax extension.
         layer : str
-            Name of the layer.
+            Name of the layer. to which apply the reference.
```
Comment (PipKat) [inline pyaedt/hfss3dlayout.py:360]:
```suggestion
            Name of the layer to apply the reference to.
```

File: pyaedt/hfss3dlayout.py [hunk 2]
```diff
@@ -356,7 +357,7 @@ def create_coax_port(self, vianame, radial_extent, layer, alignment="lower"):
...
         alignment : str, optional
             Port alignment on Layer.
```
Comment (PipKat) [inline pyaedt/hfss3dlayout.py:362]:
```suggestion
            Port alignment on the layer.
```

File: pyaedt/hfss3dlayout.py [hunk 3]
```diff
@@ -1726,3 +1728,24 @@ def _update_port_info(self, port):
         for prop in propnames:
             props[prop] = self.oeditor.GetPropertyValue("EM Design", "Excitations:{}".format(port), prop)
         return BoundaryObject3dLayout(self, port, props, "Port")
+
+    @pyaedt_function_handler()
+    def get_model_from_mesh_results(self):
+        """Search for parasolid file in results folder and provide the path.
```
Comment (PipKat) [inline pyaedt/hfss3dlayout.py:1734]:
```suggestion
        """Gets the path for the parasolid file in the results folder.
```

File: pyaedt/hfss3dlayout.py [hunk 4]
```diff
@@ -1726,3 +1728,24 @@ def _update_port_info(self, port):
...
+        Parasolid file is generated after the mesh is created in 3D Layout.
+
+        Returns
+        -------
+        str
```
Comment (PipKat) [inline pyaedt/hfss3dlayout.py:1739]:
```suggestion
        str
            Path for the parasolid file in the results folder.
```

File: pyaedt/maxwell.py [hunk 5]
```diff
@@ -1765,6 +1765,24 @@ def _create_boundary(self, name, props, boundary_type):
         self.logger.error("Error in boundary creation for %s %s.", boundary_type, name)
         return result
 
+    def get_conduction_paths(self):
+        """Get a dictionary of all conduction paths with relative objects. It works from AEDT 23R1
```
Comment (PipKat) [inline pyaedt/maxwell.py:1769]:
```suggestion
        """Get a dictionary of all conduction paths with relative objects. This method is available in AEDT 2023 R1 and later.
```

File: pyaedt/maxwell.py [hunk 6]
```diff
@@ -1765,6 +1765,24 @@ def _create_boundary(self, name, props, boundary_type):
...
+
+        Returns
+        -------
+        dict
```
Comment (PipKat) [inline pyaedt/maxwell.py:1773]:
```suggestion
        dict
            Dictionary of all conduction paths with relative objects.
```

File: pyaedt/modeler/object3dlayout.py [hunk 7]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
         ]
         return self.change_property(args)
 
+    @property
+    def solderball_enabled(self):
+        """Check if Solderball is enabled.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:353]:
```suggestion
        """Check if solderball is enabled.
```

File: pyaedt/modeler/object3dlayout.py [hunk 8]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+
+        Returns
+        -------
+        bool
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:357]:
```suggestion
        bool
            ````True`` when successful, ``False"" when failed.
```

File: pyaedt/modeler/object3dlayout.py [hunk 9]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        if self._part_type_id not in [0, 4, 5]:
+            return False
+        component_info = str(list(self.m_Editor.GetComponentInfo(self.name))).replace("'", "").replace('"', "")
+        if "sbsh=Cyl" in component_info or "sbsh=Sph" in component_info:
+            return True
+        return False
+
+    @property
+    def die_enabled(self):
+        """Check if the die is enabled. Valid for IC only.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:368]:
```suggestion
        """Check if the die is enabled. This method is valid for integrated circuits only.
```

File: pyaedt/modeler/object3dlayout.py [hunk 10]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+
+        Returns
+        -------
+        bool
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:372]:
```suggestion
        bool
            ``True`` when successful, ``False`` when failed.
```

File: pyaedt/modeler/object3dlayout.py [hunk 11]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        if self._part_type_id not in [0, 4, 5]:
+            return False
+        component_info = str(list(self.m_Editor.GetComponentInfo(self.name))).replace("'", "").replace('"', "")
+        if "dt=1" in component_info or "dt=2" in component_info:
+            return True
+        return False
+
+    @property
+    def die_type(self):
+        """Get the die type.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:383]:
```suggestion
        """Die type.
```

File: pyaedt/modeler/object3dlayout.py [hunk 12]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        self,
+        die_type=1,
+        orientation=0,
+        height=0,
+        reference_offset=0,
+        auto_reference=True,
+        reference_x="0.1mm",
+        reference_y="0.1mm",
+    ):
+        """Set the die type
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:409]:
```suggestion
        """Set the die type.
```

File: pyaedt/modeler/object3dlayout.py [hunk 13]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+
+        Parameters
+        ----------
+        die_type : int, optional
+            Die type. It can be 0 for None, 1 for FlipChip and 2 for WireBond.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:414]:
```suggestion
            Die type. The default is ``1``. Options are ``0`` for None, ``1`` for FlipChip, and ``2`` for WireBond.
```

File: pyaedt/modeler/object3dlayout.py [hunk 14]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        orientation : int, optional
+            Die orientation. it can be 0 for Chip Top and 1 for Chip Bottom.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:416]:
```suggestion
            Die orientation. The default is ``0``. Options are ``0`` for Chip Top and ``1`` for Chip Bottom.
```

File: pyaedt/modeler/object3dlayout.py [hunk 15]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        height : float, optional
+            Die height valid for port setup.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:418]:
```suggestion
            Die height valid for port setup. The default is ``0``.
```

File: pyaedt/modeler/object3dlayout.py [hunk 16]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        reference_offset : str, float, optional
+            Port reference offset. Default is 0.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:420]:
```suggestion
            Port reference offset. The default is ``0``.
```

File: pyaedt/modeler/object3dlayout.py [hunk 17]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        auto_reference : bool, optional
+            Either if auto compute reference size. Default is `True`.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:422]:
```suggestion
            Whether to automatically compute reference size. The default is ``True``.
```

File: pyaedt/modeler/object3dlayout.py [hunk 18]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        reference_x : str, float, optional
+            Reference x size. auto_reference as to be set to False to use this parameter.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:424]:
```suggestion
            Reference x size for when ``auto_reference=False``.
```

File: pyaedt/modeler/object3dlayout.py [hunk 19]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        reference_y : str, float, optional
+            Reference y size. auto_reference as to be set to False to use this parameter.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:426]:
```suggestion
            Reference y size for when ``auto_reference=False``
```

File: pyaedt/modeler/object3dlayout.py [hunk 20]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+
+        Returns
+        -------
+        bool
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:430]:
```suggestion
        bool
            ``True`` when successful, ``False`` when failed.
```

File: pyaedt/modeler/object3dlayout.py [hunk 21]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+                4,
+            ],
+        ]
+        return self.change_property(args)
+
+    @pyaedt_function_handler()
+    def set_solderball(
+        self, solderball_type="Cyl", diameter="0.1mm", mid_diameter="0.1mm", height="0.2mm", material="solder"
+    ):
+        """Set Solderball on the active component. It applies to Component of type Other, IC and IO.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:471]:
```suggestion
        """Set solderball on the active component.
        
        The method applies to these component types: ``Other``, ``IC`` and ``IO``.
```

File: pyaedt/modeler/object3dlayout.py [hunk 22]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+
+        Parameters
+        ----------
+        solderball_type : str, optional
+            Either one of "None", "Cyl" or "Sph" values.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:476]:
```suggestion
            Solderball type. The default is ``"Cyl"``. Options are ``"None"``, ``"Cyl"``, and ``"Sph"``.
```

File: pyaedt/modeler/object3dlayout.py [hunk 23]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        diameter : str, optional
+            Ball diameter.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:478]:
```suggestion
            Ball diameter. The default is ``"0.1mm"``.
```

File: pyaedt/modeler/object3dlayout.py [hunk 24]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        mid_diameter : str, optional
+            Ball diameter.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:480]:
```suggestion
            Ball diameter. The default is ``"0.1mm"``.
```

Comment (PipKat) [reply pyaedt/modeler/object3dlayout.py:480]: The description in line 480 must differ from that in line 478. Perhaps it only needs to be changed to "Mid-ball diameter"?

File: pyaedt/modeler/object3dlayout.py [hunk 25]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        height : str, optional
+            Ball mid diameter.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:482]:
```suggestion
            Ball height.
```

File: pyaedt/modeler/object3dlayout.py [hunk 26]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+        material : str, optional
+            Ball height.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:484]:
```suggestion
            Ball material.
```

File: pyaedt/modeler/object3dlayout.py [hunk 27]
```diff
@@ -348,6 +348,224 @@ def enabled(self, status=True):
...
+
+        Returns
+        -------
+        bool
+            `True` if succeeded. `False` if fails or wrong component type.
```
Comment (PipKat) [inline pyaedt/modeler/object3dlayout.py:489]:
```suggestion
            ``True`` when successful, ``False`` when failed or the wrong component type.
```

File: pyaedt/modeler/object3dlayout.py (hunk 11 above)
Comment (PipKat) [reply pyaedt/modeler/object3dlayout.py:383]: I don't think a property needs to have a Returns section. I could be wrong though. According to Alex, it should be a noun description (no "Get" verb).

**Review by Samuelopez-ansys — APPROVED** [review]
LGTM

File: pyaedt/maxwell.py [hunk 28]
```diff
@@ -1434,8 +1437,119 @@ def assign_current_density(
...
+        output_directory : str, optional
+            Path to export. If `None` pyaedt working dir will be used.
+        setup_name : str, optional
+            Setup name. If `None` pyaedt will use nominal setup.
+        start_frequency : float, optional
+            When a float is entered the Start-Stop Frequency approach is used.
+        stop_frequency : float, optional
+            A float must be entered when the Start-Stop Frequency approach is used.
+        number_of_frequency : int, optional
+            When a number is entered the number of frequencies approach is used.
```
Comment (MaxJPRey) [inline pyaedt/maxwell.py:1527]:
```suggestion
            When a number is entered, the number of frequencies approach is used.
```

File: pyaedt/maxwell.py [hunk 29]
```diff
@@ -1434,8 +1437,119 @@ def assign_current_density(
...
+        """Export Element Based Harmonic Forces csv to file.
+
+        Parameters
+        ----------
+        output_directory : str, optional
+            Path to export. If `None` pyaedt working dir will be used.
+        setup_name : str, optional
+            Setup name. If `None` pyaedt will use nominal setup.
+        start_frequency : float, optional
+            When a float is entered the Start-Stop Frequency approach is used.
```
Comment (MaxJPRey) [inline pyaedt/maxwell.py:1523]:
```suggestion
            When a float is entered the, Start-Stop Frequency approach is used.
```

File: pyaedt/maxwell.py [hunk 30]
```diff
@@ -1434,8 +1437,119 @@ def assign_current_density(
...
+        number_of_frequency=None,
+    ):
+        """Export Element Based Harmonic Forces csv to file.
+
+        Parameters
+        ----------
+        output_directory : str, optional
+            Path to export. If `None` pyaedt working dir will be used.
+        setup_name : str, optional
+            Setup name. If `None` pyaedt will use nominal setup.
```
Comment (MaxJPRey) [inline pyaedt/maxwell.py:1521]:
```suggestion
            Setup name. If ``None`` pyaedt will use nominal setup.
```

File: pyaedt/maxwell.py [hunk 31]
```diff
@@ -1434,8 +1437,119 @@ def assign_current_density(
...
+        start_frequency=None,
+        stop_frequency=None,
+        number_of_frequency=None,
+    ):
+        """Export Element Based Harmonic Forces csv to file.
+
+        Parameters
+        ----------
+        output_directory : str, optional
+            Path to export. If `None` pyaedt working dir will be used.
```
Comment (MaxJPRey) [inline pyaedt/maxwell.py:1519]:
```suggestion
            Path to export. If ``None`` pyaedt working dir will be used.
```

**Pull request merged**